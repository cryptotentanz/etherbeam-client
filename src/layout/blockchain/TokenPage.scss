@use '../../styles/resources/sizes';

.my-tokenpage {
  &-main {
    display: flex;
    flex-wrap: wrap;
  }

  &-details {
    flex-basis: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  &-link {
    margin: 0px !important;
  }

  &-chart {
    flex-basis: 100%;
  }

  @media screen and (min-width: sizes.$breakpoint-w-s) {
    &-details {
      flex-basis: 50%;
    }

    &-chart {
      flex-basis: 50%;
    }
  }
}
