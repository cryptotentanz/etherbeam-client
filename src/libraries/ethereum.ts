import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
