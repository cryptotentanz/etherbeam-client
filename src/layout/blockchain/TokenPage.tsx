import React, { FC } from 'react'
import { Token, TransactionActionType } from '../../libraries/types'
import { faHome, faFileAlt, IconDefinition, faPlusCircle, faMinusCircle } from '@fortawesome/free-solid-svg-icons'
import {
  faGithub,
  faLinkedin,
  faFacebook,
  faReddit,
  faTwitter,
  faTelegram,
  faDiscord,
} from '@fortawesome/free-brands-svg-icons'
import { getUniswapTradeUrl } from '../../components/blockchain/helpers'
import Alert from '../../components/Alert'
import Block from '../../components/Block'
import Link from '../../components/Link'
import Loader from '../../components/Loader'
import List from '../../components/List'
import ListItem from '../../components/ListItem'
import AddressItem from '../../components/blockchain/AddressItem'
import TokenAmount from '../../components/blockchain/TokenAmount'
import TokenPriceHistory from '../../components/blockchain/TokenPriceHistory'
import TokenChart from '../../components/blockchain/TokenChart'
import TransactionActionList from '../../components/blockchain/TransactionActionList'
import './TokenPage.scss'

interface IconLinkProps {
  url: string
  icon: IconDefinition
  description: string
}

const IconLink: FC<IconLinkProps> = ({ url, icon, description }: IconLinkProps) => {
  return (
    <ListItem className="my-tokenpage-link">
      <Link href={url} icon={icon} description={description} external size="l" />
    </ListItem>
  )
}

interface TokenPageProps {
  token: Token | null
  alert?: string | null
  loading?: boolean
}

const TokenPage: FC<TokenPageProps> = ({ token, alert, loading }: TokenPageProps) => {
  const {
    hash,
    name,
    symbol,
    chartPair,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    price,
    priceHistory,
    actions,
  } = token || {}
  const filteredActions = actions?.filter((action) => action.type != TransactionActionType.Approval)

  const Links: FC = () => {
    return (
      <List horizontal>
        {website && <IconLink url={website} description="Website" icon={faHome} />}
        {whitepaper && <IconLink url={whitepaper} description="Whitepaper" icon={faFileAlt} />}
        {github && <IconLink url={github} description="Github" icon={faGithub} />}
        {linkedin && <IconLink url={linkedin} description="LinkedIn" icon={faLinkedin} />}
        {facebook && <IconLink url={facebook} description="Facebook" icon={faFacebook} />}
        {reddit && <IconLink url={reddit} description="Reddit" icon={faReddit} />}
        {twitter && <IconLink url={twitter} description="Twitter" icon={faTwitter} />}
        {telegram && <IconLink url={telegram} description="Telegram" icon={faTelegram} />}
        {discord && <IconLink url={discord} description="Discord" icon={faDiscord} />}
      </List>
    )
  }

  const TradeActions: FC = () => {
    return (
      <Block>
        {hash && (
          <List horizontal center>
            <ListItem>
              <Link
                href={getUniswapTradeUrl(null, hash)}
                icon={faPlusCircle}
                label="Buy"
                color="primary"
                size="l"
                button
                external
              />
            </ListItem>
            <ListItem>
              <Link
                href={getUniswapTradeUrl(hash, null)}
                icon={faMinusCircle}
                label="Sell"
                color="primary"
                size="l"
                button
                external
              />
            </ListItem>
          </List>
        )}
      </Block>
    )
  }

  return (
    <>
      <div>
        <Block className="my-tokenpage-main">
          <div className="my-tokenpage-details">
            <div>
              {token && <h1>{name}</h1>}
              {!token && !loading && !!alert && <h1>Unknown token</h1>}
              {token && (
                <>
                  <Links />
                  <AddressItem className="my-d-min-m" address={token} noFlag />
                  <AddressItem className="my-d-max-m" address={token} noFlag short />
                  <Block>
                    <span className="my-d-min-xs my-disabled-fg my-size-l">{symbol} 1 = </span>
                    <TokenAmount amount={price} size="l" />
                  </Block>
                  <Block>{!!priceHistory?.length && <TokenPriceHistory priceHistory={priceHistory} size="m" />}</Block>
                </>
              )}
            </div>
            {token && <TradeActions />}
          </div>
          <div className="my-tokenpage-chart my-d-min-s">{token && <TokenChart pair={chartPair} height={300} />}</div>
        </Block>

        {!!filteredActions?.length && (
          <section>
            <h3>Actions</h3>
            <TransactionActionList actions={filteredActions} />
          </section>
        )}
      </div>
      {!token && loading && <Loader type="dots" />}
      {!token && !!alert && <Alert color="negative">{alert}</Alert>}
    </>
  )
}

export default TokenPage
