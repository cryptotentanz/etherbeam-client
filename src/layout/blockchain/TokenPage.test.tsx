import * as React from 'react'
import { render } from 'enzyme'
import { createToken } from 'tests/fixtures/ethereum'

jest.mock('components/blockchain/TokenChart')

import TokenPage from './TokenPage'

describe('<TokenPage />', () => {
  const token = createToken('0x0000000000000000000000000000000000000111')

  let component: unknown

  describe('Default', () => {
    beforeAll(async () => {
      component = render(<TokenPage token={token} />)
    })

    it('should match snapshot', () => expect(component).toMatchSnapshot())
  })

  describe('Loading', () => {
    beforeAll(async () => {
      component = render(<TokenPage token={null} loading />)
    })

    it('should match snapshot', () => expect(component).toMatchSnapshot())
  })
})
