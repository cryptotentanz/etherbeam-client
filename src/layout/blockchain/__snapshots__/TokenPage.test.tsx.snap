// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenPage /> Default should match snapshot 1`] = `
<div>
  <div
    class="my-block my-tokenpage-main"
  >
    <div
      class="my-tokenpage-details"
    >
      <div>
        <h1>
          Token
        </h1>
        <ul
          class="my-list-h  "
        />
        <span
          class="my-d-min-m"
        >
          <a
            class="my-clickable  my-link my-secondary-fg"
            href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="my-addresshash"
            >
              0x0000000000000000000000000000000000000111
            </span>
          </a>
        </span>
        <span
          class="my-d-max-m"
        >
          <a
            class="my-clickable  my-link my-secondary-fg"
            href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="my-addresshash"
            >
              0x0000...0111
            </span>
          </a>
        </span>
        <div
          class="my-block "
        >
          <span
            class="my-d-min-xs my-disabled-fg my-size-l"
          >
            TKN 1 = 
          </span>
          <span
            class="my-badge my-light-bg my-size-l "
          >
            <span
              class="my-tokenamount-symbol"
            >
              ETH
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
        </div>
        <div
          class="my-block "
        >
          <ul
            class="my-list-h  "
          >
            <li
              class="my-listitem "
            >
              <span
                class="my-badge my-positive-bg my-size-m "
              >
                <span
                  class="my-difference-prefix"
                >
                  H
                </span>
                <span
                  class="my-difference-amount"
                >
                  +100%
                </span>
              </span>
            </li>
            <li
              class="my-listitem "
            >
              <span
                class="my-badge my-negative-bg my-size-m "
              >
                <span
                  class="my-difference-prefix"
                >
                  D
                </span>
                <span
                  class="my-difference-amount"
                >
                  -50%
                </span>
              </span>
            </li>
            <li
              class="my-listitem "
            >
              <span
                class="my-badge my-negative-bg my-size-m "
              >
                <span
                  class="my-difference-prefix"
                >
                  W
                </span>
                <span
                  class="my-difference-amount"
                >
                  -75%
                </span>
              </span>
            </li>
            <li
              class="my-listitem "
            >
              <span
                class="my-badge my-light-bg my-size-m "
              >
                <span
                  class="my-difference-prefix"
                >
                  M
                </span>
                <span
                  class="my-difference-amount"
                >
                  +0%
                </span>
              </span>
            </li>
            <li
              class="my-listitem "
            >
              <span
                class="my-badge my-positive-bg my-size-m "
              >
                <span
                  class="my-difference-prefix"
                >
                  Y
                </span>
                <span
                  class="my-difference-amount"
                >
                  +300%
                </span>
              </span>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="my-block "
      >
        <ul
          class="my-list-h my-list-center "
        >
          <li
            class="my-listitem "
          >
            <a
              class="my-clickable  my-link-button my-primary-bg my-size-l"
              href="https://app.uniswap.org/#/swap?outputCurrency=0x0000000000000000000000000000000000000111"
              rel="noreferrer"
              tabindex="0"
              target="_blank"
            >
              <span
                class="my-icon  "
              >
                <span
                  class=" my-alt"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-plus-circle fa-w-16 "
                    data-icon="plus-circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span
                  class="my-icon-label "
                >
                  Buy
                </span>
              </span>
            </a>
          </li>
          <li
            class="my-listitem "
          >
            <a
              class="my-clickable  my-link-button my-primary-bg my-size-l"
              href="https://app.uniswap.org/#/swap?inputCurrency=0x0000000000000000000000000000000000000111&"
              rel="noreferrer"
              tabindex="0"
              target="_blank"
            >
              <span
                class="my-icon  "
              >
                <span
                  class=" my-alt"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-minus-circle fa-w-16 "
                    data-icon="minus-circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span
                  class="my-icon-label "
                >
                  Sell
                </span>
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="my-tokenpage-chart my-d-min-s"
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`<TokenPage /> Loading should match snapshot 1`] = `
Array [
  <div>
    <div
      class="my-block my-tokenpage-main"
    >
      <div
        class="my-tokenpage-details"
      >
        <div />
      </div>
      <div
        class="my-tokenpage-chart my-d-min-s"
      />
    </div>
  </div>,
  <div
    class="my-loader"
  >
    <div
      class="my-loader-dots"
    />
  </div>,
]
`;
